### PR TITLE
metadata: add infrastructure for extracting metadata from parts

### LINKED ISSUE #1693: Develop metadata handler system
Estimate: 1 week

Currently Snapcraft requires all its metadata (snap name, version, etc.) to be specified in the snapcraft.yaml. However, lots of parts already contain such information, whether from a setup.py, XML, AppStream, etc.

I’m considering this task to only be for existing metadata (name, version, summary, etc.). Adding support for an AppStream ID etc. is a separate task

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

This PR resolves #1693 by adding the underlying infrastructure necessary to extract metadata from parts. No actual metadata extraction is done here (e.g. Appstream). That will come in follow-up PRs, and will be as simple as adding a new module to `snapcraft.extractors` with an `extract` function.

### CODE REVIEW

File: schema/snapcraft.yaml [hunk 1]
```diff
@@ -361,16 +364,30 @@ properties:
...
   - parts
+
+# Either summary/description is required, or adopt-info is required to specify
+# the part from which this metadata will be retrieved.
+anyOf:
+  - required:
+    - summary
+    - description
+  - required:
+    - adopt-info
```
Comment (sergiusens) [inline schema/snapcraft.yaml]: nice

File: snapcraft/extractors/__init__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from ._metadata import ExtractedMetadata  # noqa
+from ._errors import UnhandledFileError  # noqa
```
Comment (sergiusens) [inline snapcraft/extractors/__init__.py]: I wouldn't import this, only if it is public (which I wouldn't be against)

Comment (sergiusens) [reply snapcraft/extractors/__init__.py]: err, abort!

File: snapcraft/extractors/_metadata.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        :param str summary: Extracted summary
+        :param str description: Extracted description
+        """
+
+        self._data = {}  # type: Dict[str, str]
+
+        if summary:
+            self._data['summary'] = summary
+        if description:
+            self._data['description'] = description
```
Comment (sergiusens) [inline snapcraft/extractors/_metadata.py]: what is the story with blobs like icon and desktop file?

File: snapcraft/internal/lifecycle/_runner.py [hunk 4]
```diff
@@ -58,7 +59,7 @@ def execute(step, project_options, part_names=None):
                           over.
     :returns: A dict with the snap name, version, type and architectures.
     """
-    config = snapcraft.internal.load_config(project_options)
+    config = project_loader.load_config(project_options)
```
Comment (sergiusens) [inline snapcraft/internal/lifecycle/_runner.py]: might as well consider unimporting `snapcraft`?

Comment (sergiusens) [reply snapcraft/internal/lifecycle/_runner.py]: and also `snapcraft.internal`

File: snapcraft/internal/pluginhandler/__init__.py [hunk 5]
```diff
@@ -90,6 +97,26 @@ def __init__(self, *, plugin, part_properties, project_options,
 
         self._migrate_state_file()
 
+    def pull_state(self) -> states.PullState:
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py]: these should have a `get` in front

Comment (sergiusens) [reply snapcraft/internal/pluginhandler/__init__.py]: maybe this really merits a mixin as well? Thoughts?

File: snapcraft/internal/pluginhandler/__init__.py [hunk 6]
```diff
@@ -334,9 +372,40 @@ def mark_build_done(self):
         plugin_manifest = self.plugin.get_manifest()
         machine_manifest = self._get_machine_manifest()
 
+        # Extract any requested metadata available in the build directory,
+        # followed by the install directory (which takes precedence)
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py]: \o/

File: snapcraft/internal/pluginhandler/_metadata_extraction.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def extract_metadata(part_name: str,
+                     file_path: str) -> extractors.ExtractedMetadata:
+    if not os.path.exists(file_path):
+        raise MissingMetadataFileError(part_name, file_path)
+
+    # Iterate through each extractor module, calling the 'extract' function
+    # from it. If it raises an 'UnhandledFileError' move onto the next.
+    for _, module_name, _ in pkgutil.iter_modules(
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/_metadata_extraction.py]: this is clever, I am just afraid we might have some false positives
it would be nice to be able to infer some preference given the file name as well

File: snapcraft/tests/unit/project_loader/test_config.py [hunk 8]
```diff
@@ -1916,9 +1893,11 @@ def setUp(self):
             'version': '1.0-snapcraft1~ppa1',
             'summary': 'my summary less that 79 chars',
             'description': 'description which can be pretty long',
+            'adopt-info': 'part1',
```
Comment (sergiusens) [inline snapcraft/tests/unit/project_loader/test_config.py]: why is this necessary?

File: snapcraft/tests/unit/project_loader/test_config.py [hunk 9]
```diff
@@ -2038,6 +2017,20 @@ def test_both_snap_and_prime_specified(self):
...
+    def test_missing_required_property_and_missing_adopt_info(self):
+        del self.data['summary']
+        del self.data['adopt-info']
+
+        raised = self.assertRaises(
+            errors.YamlValidationError,
+            project_loader.Validator(self.data).validate)
+
+        expected_message = (
+            "'adopt-info' is a required property or 'summary' is a required "
```
Comment (sergiusens) [inline snapcraft/tests/unit/project_loader/test_config.py]: now that I see the message here, it feels rather off.

File: snapcraft/tests/unit/test_errors.py [hunk 10]
```diff
@@ -286,7 +287,51 @@ class ErrorFormattingTestCase(unit.TestCase):
             'exception': errors.InvalidContainerImageInfoError,
             'kwargs': {'image_info': 'test-image-info'},
             'expected_message': (
-                'Error parsing the container image info: test-image-info')})
+                'Error parsing the container image info: test-image-info')}),
+        # meta errors.
```
Comment (sergiusens) [inline snapcraft/tests/unit/test_errors.py]: nice but are we bundling ALL errors here? This is just rhetoric ;-)

File: snapcraft/tests/unit/project_loader/test_config.py (hunk 9 above)
Comment (kyrofa) [reply snapcraft/tests/unit/project_loader/test_config.py]: Mannnn, this took me hours. Would you prefer what jsonschema gives us?

> Issues while validating snapcraft.yaml: OrderedDict([('name', 'my-snap-name'), ('version', '0.1'), ('description', "This is my-snap's description. You have a paragraph or two to tell the\nmost important story about your snap. Keep it under 100 words though,\nwe live in tweetspace and your description wants to look good in the snap\nstore.\n"), ('grade', 'devel'), ('confinement', 'devmode'), ('parts', OrderedDict([('my-part', OrderedDict([('plugin', 'nil'), ('parse-info', ['foo'])]))]))]) is not valid under any of the given schemas

Just kidding. I'd love to improve it, but couldn't figure out how to do so in a generic manner. Should I hard-code handling for this specific case?

File: snapcraft/tests/unit/project_loader/test_config.py (hunk 8 above)
Comment (kyrofa) [reply snapcraft/tests/unit/project_loader/test_config.py]: Just so I don't have to explicitly add it in the test cases. Harmless for the other ones.

File: snapcraft/internal/pluginhandler/_metadata_extraction.py (hunk 7 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/_metadata_extraction.py]: Yeah I was thinking about preference as well, although I couldn't think of any extractors that overlap-- do we ever anticipate that multiple extractors could handle the same file? If not, no preference is needed.

File: snapcraft/extractors/_metadata.py (hunk 3 above)
Comment (kyrofa) [reply snapcraft/extractors/_metadata.py]: As we briefly discussed, we can base64 them, or save file paths here and grab them when creating the meta/ dir.

File: snapcraft/internal/pluginhandler/__init__.py (hunk 5 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/__init__.py]: Oh right-- I'll fix the lack of "get". Regarding the mixin, I feel it's only worth it if we want the functionality mixed into another class as well. I don't think we do... do we?

File: snapcraft/internal/lifecycle/_runner.py (hunk 4 above)
Comment (kyrofa) [reply snapcraft/internal/lifecycle/_runner.py]: We use `snapcraft`, but you're right, `snapcraft.internal` does not seem necessary. Wonder why flake8 didn't grab that. Fixed.

File: snapcraft/internal/pluginhandler/__init__.py (hunk 5 above)
Comment (kyrofa) [reply snapcraft/internal/pluginhandler/__init__.py]: Fixed.
